packages/muelu/src/Misc/MueLu_SegregatedAFactory_def.hpp

```diff
@@ -55,114 +55,273 @@
-      for(size_t i=0; i<(size_t)indices.size(); i++) { // or can be parallelize this loop?
-        GlobalOrdinal gcid = Ain->getColMap()->getGlobalElement(indices[i]); // global column id
-        bool isInMap2 = map->isNodeGlobalElement(gcid);
-
-        if (isInMap == isInMap2) {
-          indout [nNonzeros] = gcid;
-          valout [nNonzeros] = vals[i];
-          nNonzeros++;
-        }
+      if (isInMap == isInMap2) {
+        indout [nNonzeros] = gcid;
+        valout [nNonzeros] = vals[i];
+        nNonzeros++;
       }
-      indout.resize(nNonzeros);
-      valout.resize(nNonzeros);
+    }
+    indout.resize(nNonzeros);
+    valout.resize(nNonzeros);
+
+    Aout->insertGlobalValues(Ain->getRowMap()->getGlobalElement(row), indout.view(0,indout.size()), valout.view(0,valout.size()));
+  }
+
+  Aout->fillComplete(Ain->getDomainMap(), Ain->getRangeMap());
+
+  // copy block size information
+  Aout->SetFixedBlockSize(Ain->GetFixedBlockSize());
+
+  GetOStream(Statistics0, 0) << "Nonzeros in A (input): " << Ain->getGlobalNumEntries() << ", Nonzeros after filtering A: " << Aout->getGlobalNumEntries() << std::endl;
+
+  Set(currentLevel, "A", Aout);
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void SegregatedAFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::BuildBasedOnMapPair(Level &currentLevel) const
+{
+  FactoryMonitor m(*this, "Matrix filtering (segregation, map-pair)", currentLevel);
+
+  RCP<Matrix> Ain = Get< RCP<Matrix> >(currentLevel, "A");
 
-      Aout->insertGlobalValues(Ain->getRowMap()->getGlobalElement(row), indout.view(0,indout.size()), valout.view(0,valout.size()));
+  const ParameterList &pL = GetParameterList();
+  auto comm = Ain->getRowMap()->getComm();
+
+  // fetch maps from level
+  RCP<const Map> dropMap1 = Teuchos::null;
+  RCP<const Map> dropMap2 = Teuchos::null;
+
+  const std::string dropMap1Name = "dropMap1";
+  const std::string dropMap2Name = "dropMap2";
+
+  if (currentLevel.GetLevelID() == 0) {
+    dropMap1 = currentLevel.Get<RCP<const Map>>(dropMap1Name, NoFactory::get());
+    dropMap2 = currentLevel.Get<RCP<const Map>>(dropMap2Name, NoFactory::get());
+    GetOStream(Statistics0) << "User provided dropMap1 \"" << dropMap1Name << "\": length dimension=" << dropMap1->getGlobalNumElements() << std::endl;
+    GetOStream(Statistics0) << "User provided dropMap2 \"" << dropMap2Name << "\": length dimension=" << dropMap2->getGlobalNumElements() << std::endl;
+  }
+  else {
+    dropMap1 = Get<RCP<const Map>>(currentLevel, dropMap1Name);
+    dropMap2 = Get<RCP<const Map>>(currentLevel, dropMap2Name);
+  }
+
+  TEUCHOS_ASSERT(!dropMap1.is_null());
+  TEUCHOS_ASSERT(!dropMap2.is_null());
+
+  // create new empty Operator
+  Teuchos::RCP<Matrix> Aout = MatrixFactory::Build(Ain->getRowMap(), Ain->getGlobalMaxNumRowEntries());
+
+  Teuchos::RCP<const Map> redundantDropMap1 = Teuchos::null;
+  Teuchos::RCP<const Map> redundantDropMap2 = Teuchos::null;
```

Comment:
Maybe rename the variables from `redundant...` to `final...`, since they can be redundant, but don't have to.

Reply:
i like it. Applied.